File: pathfinder-editmode/src/main/java/de/cubbossa/pathfinder/editmode/renderer/NodeEntityRenderer.java
package de.cubbossa.pathfinder.editmode.renderer;

import de.cubbossa.menuframework.inventory.Action;
import de.cubbossa.menuframework.inventory.context.TargetContext;
import de.cubbossa.pathfinder.PathFinderProvider;
import de.cubbossa.pathfinder.misc.PathPlayer;
import de.cubbossa.pathfinder.node.Node;
import de.cubbossa.pathfinder.util.BukkitVectorUtils;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.entity.BlockDisplay;
import org.bukkit.entity.Interaction;
import org.bukkit.entity.Player;
import org.bukkit.plugin.java.JavaPlugin;
import org.bukkit.util.Transformation;
import org.joml.Vector3f;

import java.util.Collection;
import java.util.concurrent.CompletableFuture;

import static de.cubbossa.pathfinder.editmode.menu.EditModeMenu.LEFT_CLICK_NODE;
import static de.cubbossa.pathfinder.editmode.menu.EditModeMenu.RIGHT_CLICK_NODE;

public class NodeEntityRenderer extends AbstractEntityRenderer<Node, BlockDisplay> {


  private static final float NODE_SCALE = .4f;

  public NodeEntityRenderer(JavaPlugin plugin) {
    super(plugin, BlockDisplay.class);
    setRenderDistance(PathFinderProvider.get().getConfiguration().getEditMode().getNodeArmorStandRenderDistance());
  }

  @Override
  public CompletableFuture<Void> clear(PathPlayer<Player> player) {
    return CompletableFuture.runAsync(() -> {
      hideElements(entityNodeMap.values(), player.unwrap());
      players.remove(player);
    }).exceptionally(throwable -> {
      throwable.printStackTrace();
      return null;
    });
  }

  @Override
  public CompletableFuture<Void> renderNodes(PathPlayer<Player> player, Collection<Node> nodes) {
    return CompletableFuture.runAsync(() -> {
      showElements(nodes, player.unwrap()).join();
      players.add(player);
    }).exceptionally(throwable -> {
      throwable.printStackTrace();
      return null;
    });
  }

  @Override
  public CompletableFuture<Void> eraseNodes(PathPlayer<Player> player, Collection<Node> nodes) {
    return CompletableFuture.runAsync(() -> {
      hideElements(nodes, player.unwrap());
    }).exceptionally(throwable -> {
      throwable.printStackTrace();
      return null;
    });
  }

  @Override
  boolean equals(Node a, Node b) {
    return a.equals(b);
  }

  @Override
  CompletableFuture<Location> location(Node element) {
    return CompletableFuture.completedFuture(BukkitVectorUtils.toBukkit(element.getLocation()));
  }

  @Override
  Action<TargetContext<Node>> handleInteract(Player player, int slot, boolean left) {
    return left ? LEFT_CLICK_NODE : RIGHT_CLICK_NODE;
  }

  @Override
  void render(Node element, BlockDisplay entity) {
    entity.setBlock(Material.LIME_CONCRETE.createBlockData());
    Transformation t = entity.getTransformation();
    t.getTranslation().sub(new Vector3f(NODE_SCALE, NODE_SCALE, NODE_SCALE).mul(0.5f));
    t.getScale().set(NODE_SCALE);
    entity.setTransformation(t);
  }

  @Override
  void hitbox(Node element, Interaction entity) {
    entity.setInteractionWidth(NODE_SCALE);
    entity.setInteractionHeight(NODE_SCALE);
    entity.teleport(entity.getLocation().subtract(0, NODE_SCALE / 2., 0));
  }
}

File: pathfinder-editmode/src/main/kotlin/de/cubbossa/pathfinder/editmode/renderer/NodeEntityRenderer.kt
package de.cubbossa.pathfinder.editmode.renderer

import de.cubbossa.menuframework.inventory.Action
import de.cubbossa.menuframework.inventory.context.TargetContext
import de.cubbossa.pathfinder.PathFinder
import de.cubbossa.pathfinder.editmode.menu.LEFT_CLICK_NODE
import de.cubbossa.pathfinder.editmode.menu.RIGHT_CLICK_NODE
import de.cubbossa.pathfinder.misc.PathPlayer
import de.cubbossa.pathfinder.node.Node
import de.cubbossa.pathfinder.util.BukkitVectorUtils
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.entity.BlockDisplay
import org.bukkit.entity.Interaction
import org.bukkit.entity.Player
import org.bukkit.plugin.java.JavaPlugin
import org.joml.Vector3f

class NodeEntityRenderer(plugin: JavaPlugin) :
    AbstractEntityRenderer<Node, BlockDisplay>(plugin, BlockDisplay::class.java) {
    init {
        setRenderDistance(PathFinder.get().configuration.editMode.nodeArmorStandRenderDistance.toDouble())
    }

    override suspend fun clear(player: PathPlayer<Player>) {
        hideElements(entityNodeMap.values, player.unwrap()!!)
        players.remove(player)
    }

    override suspend fun renderNodes(player: PathPlayer<Player>, nodes: Collection<Node>) {
        showElements(nodes, player.unwrap()!!)
        players.add(player)
    }

    override suspend fun eraseNodes(player: PathPlayer<Player>, nodes: Collection<Node>) {
        hideElements(nodes, player.unwrap())
    }

    override fun equals(a: Node, b: Node): Boolean {
        return a == b
    }

    override suspend fun location(element: Node): Location {
        return BukkitVectorUtils.toBukkit(element.location)
    }

    override fun handleInteract(
        player: Player,
        slot: Int,
        left: Boolean
    ): Action<TargetContext<Node>> {
        return if (left) LEFT_CLICK_NODE else RIGHT_CLICK_NODE
    }

    override fun render(element: Node, entity: BlockDisplay) {
        entity.block = Material.LIME_CONCRETE.createBlockData()
        val t = entity.transformation
        t.translation.sub(Vector3f(NODE_SCALE, NODE_SCALE, NODE_SCALE).mul(0.5f))
        t.scale.set(NODE_SCALE)
        entity.transformation = t
    }

    override fun hitbox(element: Node, entity: Interaction) {
        entity.interactionWidth = NODE_SCALE
        entity.interactionHeight = NODE_SCALE
        entity.teleport(entity.location.subtract(0.0, NODE_SCALE / 2.0, 0.0))
    }

    companion object {
        private const val NODE_SCALE = .4f
    }
}
